File: Leetcode/104.二叉树的最大深度.cpp
/*
 * @lc app=leetcode.cn id=104 lang=cpp
 *
 * [104] 二叉树的最大深度
 */
#include <algorithm>
#include <iostream>
#include <vector>
#include <queue>

using namespace std;
/*2021-09-16-17-25*/
// ? 思路解析：
/*
 *
 *
 !
 !
 */
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right)
        : val(x), left(left), right(right) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */

class Solution_循环写法 {
   public:
    int maxDepth(TreeNode *root) {
        if (root == NULL) return 0;
        int depth = 0;
        queue<TreeNode *> que;
        que.push(root);

        while (!que.empty()) {
            int size = que.size();
            depth++;  // 记录深度
            for (int i = 0; i < size; i++) {
                TreeNode *node = que.front();
                que.pop();
                if (node->left) que.push(node->left);
                if (node->right) que.push(node->right);
            }
        }
        return depth;
    }
};

// @lc code=end

File: Leetcode/104.二叉树的最大深度.go
/*
 * @lc app=leetcode.cn id=104 lang=golang
 *
 * [104] 二叉树的最大深度
 */
package leetcode

/* --- 2021-10-26-17-08 --- */

// https://github.com/fengwei2002/Algorithm
// solution link: 
// https://leetcode-cn.com/problems/maximum-depth-of-binary-tree/solution/lc104-fengwei2002-by-kycu-7p9e/

type TreeNode struct {
	Val int
	Left *TreeNode
	Right *TreeNode
}

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxDepth(root *TreeNode) int {
	ans := 0
	if root == nil {
		return ans
	}
	ans = dfs(root)
	return ans
}

func dfs(root *TreeNode) int {
	if root.Left == nil && root.Right == nil {
		return 1
	}
	if root.Left != nil && root.Right == nil {
		return dfs(root.Left) + 1
	}
	if root.Left == nil && root.Right != nil {
		return dfs(root.Right) + 1
	}
	return max(dfs(root.Left), dfs(root.Right)) + 1
}

func max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}
// @lc code=end

